pants-plugins/stevedore_extensions/python_target_dependencies.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2023 The StackStorm Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import List, Mapping, Tuple
+
+from pants.backend.python.target_types import (
+    PythonTestTarget,
+    PythonTestsGeneratorTarget,
+    PythonTestsDependenciesField,
+)
+from pants.engine.addresses import Address
+from pants.engine.rules import collect_rules, rule, UnionRule
+from pants.engine.target import (
+    AllTargets,
+    FieldSet,
+    InferDependenciesRequest,
+    InferredDependencies,
+)
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import OrderedSet
+
+from stevedore_extensions.target_types import (
+    AllStevedoreExtensionTargets,
+    StevedoreEntryPointsField,
+    StevedoreExtension,
+    StevedoreNamespaceField,
+    StevedoreNamespacesField,
+)
+
+
+# -----------------------------------------------------------------------------------------------
+# Utility rules to analyze all `StevedoreExtension` targets
+# -----------------------------------------------------------------------------------------------
+
+
+@rule(desc="Find all StevedoreExtension targets in project", level=LogLevel.DEBUG)
+def find_all_stevedore_extension_targets(
+    targets: AllTargets,
+) -> AllStevedoreExtensionTargets:
+    return AllStevedoreExtensionTargets(
+        tgt for tgt in targets if tgt.has_field(StevedoreEntryPointsField)
+    )
+
+
+@dataclass(frozen=True)
+class StevedoreExtensions:
+    """A mapping of stevedore namespaces to a list the targets that provide them"""
```

Comment:
Doh! That clearly should have had an "of".

The simple fix:
```suggestion
    """A mapping of stevedore namespaces to a list of targets that provide them"""
```

Or would something like this be better:
```suggestion
    """A mapping of StevedoreExtension targets indexed by the namespaces they provide."""
```

I don't like the name `StevedoreExtensions` like this because the name doesn't hint at when to use it versus using `AllStevedoreExtensionTargets`. Maybe this should actually be renamed from `StevedoreExtensions` to `StevedoreExtensionsIndex` or `StevedoreExtensionsByNamespace`. Do any of these make the purpose clearer?

Reply:
I applied your suggestion. Thanks!